### PR TITLE
[draft, not for merge] testing frozen trunk2, which tracks values from trunk

### PR DESCRIPTION
**Pull request by min-xu-ai**
This also uncovers a bug where if requires_grad is set before fsdp wrapping, it is not working correctly. Will need to be fixed later.

## What does this PR do?
Fixes # (issue).

## Before submitting

- [ ] Did you have fun?
  - Make sure you had fun coding 🙃
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: tests/nn/data_parallel/test_fsdp_freezing_weights.py [hunk 1]
```diff
@@ -65,6 +68,7 @@ def _create_model(with_fsdp, with_nested_trunk):
         model = Model()
         if with_fsdp:
             model.trunk = FSDP(model.trunk)
+            model.trunk2 = FSDP(model.trunk2)
```
Comment (min-xu-ai) [inline tests/nn/data_parallel/test_fsdp_freezing_weights.py:71]: do wrapping here.

File: tests/nn/data_parallel/test_fsdp_freezing_weights.py [hunk 2]
```diff
@@ -99,10 +103,14 @@ def _distributed_worker(
     model = model.cuda()
 
     # freezing the trunk using requires_grad.
-    if freezing_method == FreezingMethod.RequiresGrad:
+    if False and freezing_method == FreezingMethod.RequiresGrad:
         for param in model.trunk.parameters():
             param.requires_grad = False
 
+    # Freeze trunk2.
+    for param in model.trunk2.parameters():
```
Comment (min-xu-ai) [inline tests/nn/data_parallel/test_fsdp_freezing_weights.py:111]: freeze after wrapped.

File: tests/nn/data_parallel/test_fsdp_freezing_weights.py [hunk 3]
```diff
@@ -121,11 +129,21 @@ def _distributed_worker(
...
         optimizer.zero_grad()
         fake_loss.backward()
-        if freezing_method == FreezingMethod.GradToNone:
+        if False and freezing_method == FreezingMethod.GradToNone:
             for param in model.trunk.parameters():
                 param.grad = None
         optimizer.step()
 
+        # Now let's update trunk2
+        for param, tracked_param in zip(
```
Comment (min-xu-ai) [inline tests/nn/data_parallel/test_fsdp_freezing_weights.py:138]: updating no need to use summon_full_params

File: tests/nn/data_parallel/test_fsdp_freezing_weights.py [hunk 4]
```diff
@@ -136,6 +154,16 @@ def _distributed_worker(
         state_after = model.module.cpu().state_dict()
         torch.save(state_after, rank_0_output)
 
+    # trunk2 is already checked above with state_dict, but let's double check with
+    # summon_full_params() here.
+    params=[]
+    context = model.module.trunk2.summon_full_params() if with_fsdp else contextlib.suppress()
```
Comment (min-xu-ai) [inline tests/nn/data_parallel/test_fsdp_freezing_weights.py:160]: here we use summon to double check

File: tests/nn/data_parallel/test_fsdp_freezing_weights.py (hunk 2 above)
Comment (tmarkstrum) [reply tests/nn/data_parallel/test_fsdp_freezing_weights.py:111]: For education purpose, what does freeze mean? What is the purpose of freeze?

Comment (min-xu-ai) [reply tests/nn/data_parallel/test_fsdp_freezing_weights.py:111]: Very good question. Thanks for asking!

In neural network training, the parameters are updated (optimized) by learning from the data and the losses. Freezing means we want to keep certain parameters fixed without them being updated or optimized by this process. This allows the rest of the parameters to learn and keep some part of the model fixed. This is a common thing to do in training.

**Comment by min-xu-ai** [comment]
debugging PR. closing.

**Pull request closed without merging**